File: api/src/main/java/com/sourceforgery/swedbank/CookieMonster.java
package com.sourceforgery.swedbank;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.HttpUrl;

public class CookieMonster implements CookieJar {

    private Set<CookieId> cookieJar = new LinkedHashSet<>();


    public void saveFromResponse(final HttpUrl url, final List<Cookie> cookies) {
        for (Cookie cookie : cookies) {
            CookieId o = new CookieId(cookie);
            cookieJar.remove(o);
            cookieJar.add(o);
        }
    }

    public List<Cookie> loadForRequest(final HttpUrl url) {
        return cookieJar.stream()
                        .filter(e -> e.cookie.matches(url))
                        .map(CookieId::getCookie)
                        .collect(Collectors.toList());
    }

    private static class CookieId {
        private final Cookie cookie;

        private CookieId(final Cookie cookie) {
            this.cookie = cookie;
        }

        @Override
        public boolean equals(Object other) {
            if (other == null) {
                return false;
            }
            if (other.getClass() != getClass()) {
                return false;
            }
            CookieId that = (CookieId) other;
            return that.cookie.name().equals(this.cookie.name())
                    && that.cookie.domain().equals(this.cookie.domain())
                    && that.cookie.path().equals(this.cookie.path());
        }

        @Override
        public int hashCode() {
            int hash = 17;
            hash = 31 * hash + cookie.name().hashCode();
            hash = 31 * hash + cookie.domain().hashCode();
            hash = 31 * hash + cookie.path().hashCode();
            return hash;
        }

        @Override
        public String toString() {
            return cookie.toString();
        }

        Cookie getCookie() {
            return cookie;
        }
    }
}

File: api/src/main/java/com/sourceforgery/swedbank/CookieMonster.kt
package com.sourceforgery.swedbank

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl

class CookieMonster : CookieJar {

    val cookieJar = LinkedHashSet<CookieId>()
        get() = LinkedHashSet(field)


    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        for (cookie in cookies) {
            val o = CookieId(cookie)
            cookieJar.remove(o)
            cookieJar.add(o)
        }
    }

    fun serializeAllCookies(): String {
        return Gson().toJson(cookieJar)
    }

    fun deserializeAllCookies(cookies: Set<CookieId>) {
        cookieJar.clear()
        cookieJar.addAll(cookies)
    }

    override fun loadForRequest(url: HttpUrl): List<Cookie> {
        return cookieJar
                .filter { e -> e.cookie.matches(url) }
                .map { it.cookie }
                .toList()
    }

    class CookieId constructor(internal val cookie: Cookie) {

        override fun equals(other: Any?): Boolean {
            if (other == null) {
                return false
            }
            if (other.javaClass != javaClass) {
                return false
            }
            val that = other as CookieId?
            return that!!.cookie.name() == this.cookie.name()
                    && that.cookie.domain() == this.cookie.domain()
                    && that.cookie.path() == this.cookie.path()
        }

        override fun hashCode(): Int {
            var hash = 17
            hash = 31 * hash + cookie.name().hashCode()
            hash = 31 * hash + cookie.domain().hashCode()
            hash = 31 * hash + cookie.path().hashCode()
            return hash
        }

        override fun toString(): String {
            return cookie.toString()
        }
    }
}
